src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -3231,8 +3253,9 @@ private void BindArrayInitializerExpressions(InitializerExpressionSyntax initial
             }
 
             ArrayTypeSyntax arrayTypeSyntax = (ArrayTypeSyntax)typeSyntax;
-            TypeSyntax elementTypeSyntax = arrayTypeSyntax.ElementType;
-            var elementType = BindType(elementTypeSyntax, diagnostics);
+            var elementTypeSyntax = arrayTypeSyntax.ElementType;
+            var arrayType = (ArrayTypeSymbol)BindArrayType(arrayTypeSyntax, diagnostics, permitDimensions: true, basesBeingResolved: null).TypeSymbol;
```

Comment:
>BindArrayType [](start = 45, length = 13)

I think this change is the cause of [https://github.com/dotnet/roslyn/issues/33945](https://github.com/dotnet/roslyn/issues/33945)
We didn't use to check whether the type is `IsRestrictedType`, but now we do (inside `BindArrayType`).